File: output/Race2012/Components/AJAXPortalModuleBase.cs
/*
' Copyright (c) 2014 Crewstone Consulting ltd
'  All rights reserved.
' 
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
' THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
' CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
' DEALINGS IN THE SOFTWARE.
' 
*/
using System;
using System.Collections.Generic;
using System.Web.Script.Serialization;
using System.Web.UI;
using DotNetNuke.Entities.Modules;
using DotNetNuke.UI.Utilities;

namespace NameSpace.Modules.Race2012
{
    public class AjaxPortalModuleBase: PortalModuleBase, IScriptControl
    {
        public delegate void AddScriptComponentDescriptorsDelegate(ScriptComponentDescriptor Descriptor);
        public delegate void AddScriptReferencesDelegate(List<ScriptReference> References);
        public delegate void AddLocalizedMessagesDelegate(Dictionary<string, string> Messages);
        protected event AddScriptComponentDescriptorsDelegate AddScriptComponentDescriptors;
        protected event AddScriptReferencesDelegate AddScriptReferences;
        protected event AddLocalizedMessagesDelegate AddLocalizedMessages;

#region Event Handlers
        //Enable Control Callbacks for this module
        //We are passing in the second argument to ensure our callbacks reach this instance of our object
        private void Page_Init(object sender, System.EventArgs e) 
        {
            ClientAPI.RegisterControlMethods(this, this.ClientID);
        }

        private void Page_PreRender(object sender, System.EventArgs e)
        {
            //Tell DotNetNuke we need to use the ScriptManager
            DotNetNuke.Framework.AJAX.RegisterScriptManager();

            //Register our client-side Script Control
            ScriptManager.GetCurrent(this.Page).RegisterScriptControl(this);
            ScriptManager.GetCurrent(this.Page).RegisterScriptDescriptors(this);
        }

#endregion

#region IScriptControl
        public IEnumerable<ScriptDescriptor> GetScriptDescriptors() 
        {
            List<ScriptDescriptor> descs = new List<ScriptDescriptor>();
            ScriptComponentDescriptor desc = new ScriptComponentDescriptor("ClientNamespaceHere");
            JavaScriptSerializer jss = new JavaScriptSerializer();
            Dictionary<string, string> msgs = new Dictionary<string, string>();

            desc.AddScriptProperty("id", String.Format("'{0}'", this.ClientID));

            if (AddLocalizedMessages != null)
                AddLocalizedMessages(msgs);

            desc.AddScriptProperty("msgs", jss.Serialize(msgs)); //Note:  should guard against messages with '

            if (AddScriptComponentDescriptors != null)
                AddScriptComponentDescriptors(desc);

            descs.Add(desc);
            return descs;
        }

        public IEnumerable<ScriptReference> GetScriptReferences() 
        {
            List<ScriptReference> refs = new List<ScriptReference>();
            if (AddScriptReferences != null)
                AddScriptReferences(refs);
            return refs;
        }

#endregion

    }
}

File: output/Race2012/Components/AJAXPortalModuleBase.vb
Imports System.Web.Script.Serialization
Imports DotNetNuke
Imports DotNetNuke.Common
Imports DotNetNuke.UI.Utilities

Namespace Crewstone Consulting ltd.Modules.Race2012

    Public Class AjaxPortalModuleBase
        Inherits Entities.Modules.PortalModuleBase : Implements IScriptControl

        Public Event AddScriptComponentDescriptors(ByVal Descriptor As ScriptComponentDescriptor)
        Public Event AddScriptReferences(ByVal References As List(Of ScriptReference))
        Public Event AddLocalizedMessages(ByVal Messages As Dictionary(Of String, String))

#Region "Event Handlers"
        'Enable Control Callbacks for this module
        'We are passing in the second argument to ensure our callbacks reach this instance of our object
        Private Sub Page_Init(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Init
            ClientAPI.RegisterControlMethods(Me, Me.ClientID)
        End Sub

        Private Sub Page_PreRender(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.PreRender
            'Tell DotNetNuke we need to use the ScriptManager
            DotNetNuke.Framework.AJAX.RegisterScriptManager()

            'Register our client-side Script Control
            ScriptManager.GetCurrent(Me.Page).RegisterScriptControl(Me)
            ScriptManager.GetCurrent(Me.Page).RegisterScriptDescriptors(Me)
        End Sub

#End Region

#Region "IScriptControl"
        Private Function GetScriptDescriptors() As IEnumerable(Of ScriptDescriptor) Implements IScriptControl.GetScriptDescriptors
            Dim descs As List(Of ScriptDescriptor) = New List(Of ScriptDescriptor)
            Dim desc As ScriptComponentDescriptor = New ScriptComponentDescriptor("ClientNamespaceHere")
            Dim jss As JavaScriptSerializer = New JavaScriptSerializer()
            Dim msgs As Dictionary(Of String, String) = New Dictionary(Of String, String)

            desc.AddScriptProperty("id", String.Format("'{0}'", Me.ClientID))

            RaiseEvent AddLocalizedMessages(msgs)
            desc.AddScriptProperty("msgs", String.Format("'{0}'", jss.Serialize(msgs))) 'Note:  should guard against messages with '

            RaiseEvent AddScriptComponentDescriptors(desc)

            descs.Add(desc)
            Return descs
        End Function

        Private Function GetScriptReferences() As IEnumerable(Of ScriptReference) Implements IScriptControl.GetScriptReferences
            Dim refs As List(Of ScriptReference) = New List(Of ScriptReference)
            RaiseEvent AddScriptReferences(refs)
            Return refs
        End Function

#End Region

    End Class

End Namespace